### ISSUE TITLE
About fp16 training

### ISSUE THREAD

**Issue by Ericzhang1** [aqlaboratory/openfold#69]
When the model is put in fp16 mode, does that mean all operations are in fp16? Also, is it possible to switch to a different optimization tool, say apex?

**Comment by gahdritz** [comment]
FP16 uses torch mixed precision. Under this scheme, a select subset of operations (e.g. softmax) are run in full precision, while other operations (e.g. matmul) are automatically cast to half precision. For more information, see https://pytorch.org/docs/stable/amp.html. If you're using DeepSpeed, which uses its own AMP implementation, you have the option to specify apex-like "levels" of mixed precision in the DeepSpeed config. See https://www.deepspeed.ai/features/#mixed-precision-training.

**Comment by gahdritz** [comment]
BTW: I just learned that Pytorch-Lightning has an `--amp_backend` flag that lets you switch to apex if you want even without DeepSpeed.
